File: app/src/org/commcare/fragments/personalId/PersonalIdBackupCodeFragment.java
package org.commcare.fragments.personalId;

import android.app.Activity;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;

import org.commcare.CommCareNoficationManager;
import org.commcare.activities.connect.viewmodel.PersonalIdSessionDataViewModel;
import org.commcare.android.database.connect.models.ConnectUserRecord;
import org.commcare.android.database.connect.models.PersonalIdSessionData;
import org.commcare.connect.ConnectConstants;
import org.commcare.connect.ReleaseToggleHelper;
import org.commcare.connect.database.ConnectDatabaseHelper;
import org.commcare.connect.database.ConnectUserDatabaseUtil;
import org.commcare.connect.network.PersonalIdOrConnectApiErrorHandler;
import org.commcare.connect.network.connectId.PersonalIdApiHandler;
import org.commcare.dalvik.R;
import org.commcare.dalvik.databinding.FragmentRecoveryCodeBinding;
import org.commcare.google.services.analytics.AnalyticsParamValue;
import org.commcare.google.services.analytics.FirebaseAnalyticsUtil;
import org.commcare.personalId.PersonalIdPreferences;
import org.commcare.personalId.PersonalIdRecoveryCompleter;
import org.commcare.utils.MediaUtil;
import org.commcare.utils.NotificationUtil;
import org.commcare.views.connect.NumericCodeView;
import org.javarosa.core.model.utils.DateUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Date;

public class PersonalIdBackupCodeFragment extends BasePersonalIdFragment {
    private static final int BACKUP_CODE_LENGTH = 6; //Note: This is brittle, defined in several places
    private FragmentRecoveryCodeBinding binding;
    private Activity activity;
    private boolean isRecovery = false;
    private int titleId;
    private PersonalIdSessionData personalIdSessionData;

    @Override
    public void onResume() {
        super.onResume();
        validateBackupCodeInputs();
        binding.backupCodeView.requestFocus(activity);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentRecoveryCodeBinding.inflate(inflater, container, false);
        activity = requireActivity();
        personalIdSessionData = new ViewModelProvider(requireActivity()).get(
                PersonalIdSessionDataViewModel.class).getPersonalIdSessionData();
        configureUiByMode();
        setupListeners();
        clearBackupCodeFields();
        activity.setTitle(getString(titleId));
        return binding.getRoot();
    }

    private void configureUiByMode() {
        isRecovery = personalIdSessionData.getAccountExists();
        if (isRecovery) {
            titleId = R.string.connect_backup_code_title_confirm;
            binding.recoveryCodeTilte.setText(R.string.connect_backup_code_message_title);
            binding.backupCodeSubtitle.setText(R.string.connect_backup_code_message);
            binding.backupCodeLayout.setVisibility(View.VISIBLE);
            binding.confirmCodeLabel.setVisibility(View.GONE);
            binding.confirmCodeLayout.setVisibility(View.GONE);
            setUserNameAndPhoto();
        } else {
            titleId = R.string.connect_backup_code_title_set;
            binding.backupCodeSubtitle.setText(getString(R.string.connect_backup_code_remember, BACKUP_CODE_LENGTH));
            binding.backupCodeLayout.setVisibility(View.VISIBLE);
            binding.confirmCodeLabel.setVisibility(View.VISIBLE);
            binding.confirmCodeLayout.setVisibility(View.VISIBLE);
            binding.welcomeBackLayout.setVisibility(View.GONE);
        }
    }

    private void setUserNameAndPhoto() {
        binding.welcomeBack.setText(getString(R.string.personalid_welcome_back_msg, personalIdSessionData.getUserName()));
        if (!TextUtils.isEmpty(personalIdSessionData.getPhotoBase64())) {
            binding.userPhoto.setImageBitmap(
                    MediaUtil.decodeBase64EncodedBitmap(personalIdSessionData.getPhotoBase64()));
        }
    }

    private void setupListeners() {
        NumericCodeView.OnCodeChangedListener codeChangedListener = code -> validateBackupCodeInputs();

        binding.backupCodeView.setOnCodeChangedListener(codeChangedListener);
        binding.confirmCodeView.setOnCodeChangedListener(codeChangedListener);

        binding.backupCodeView.setCodeCompleteListener(code -> {
            if (isRecovery && binding.connectBackupCodeButton.isEnabled()) {
                handleBackupCodeSubmission();
            }
        });

        binding.confirmCodeView.setCodeCompleteListener(code -> {
            if (binding.connectBackupCodeButton.isEnabled()) {
                handleBackupCodeSubmission();
            }
        });

        NumericCodeView.OnEnterKeyPressedListener enterKeyListener = () -> {
            if (binding.connectBackupCodeButton.isEnabled()) {
                handleBackupCodeSubmission();
            }
        };
        binding.backupCodeView.setOnEnterKeyPressedListener(enterKeyListener);
        binding.confirmCodeView.setOnEnterKeyPressedListener(enterKeyListener);

        binding.connectBackupCodeButton.setOnClickListener(v -> handleBackupCodeSubmission());
        binding.notMeButton.setOnClickListener(v -> handleNotMeButtonPressed());

        binding.backupCodeVisibilityToggle.setOnClickListener(v -> togglePasswordVisibility(
                binding.backupCodeView, binding.backupCodeVisibilityToggle));
        binding.confirmCodeVisibilityToggle.setOnClickListener(v -> togglePasswordVisibility(
                binding.confirmCodeView, binding.confirmCodeVisibilityToggle));
    }

    private void togglePasswordVisibility(NumericCodeView codeView, ImageView toggle) {
        boolean newVisibilityState = !codeView.isPasswordVisible();
        codeView.setPasswordVisible(newVisibilityState);
        toggle.setImageResource(newVisibilityState
                ? R.drawable.ic_visibility_off_24
                : R.drawable.ic_visibility_24);
    }

    private void clearBackupCodeFields() {
        binding.confirmCodeView.clearCode();
        binding.backupCodeView.clearCode();
    }

    private void validateBackupCodeInputs() {
        String backupCode1 = binding.backupCodeView.getCodeValue();
        String backupCode2 = binding.confirmCodeView.getCodeValue();

        String errorText = "";
        boolean isValid = false;

        if (backupCode1.length() == BACKUP_CODE_LENGTH) {
            if (!isRecovery && !backupCode1.equals(backupCode2)) {
                if (backupCode2.length() == BACKUP_CODE_LENGTH) {
                    errorText = getString(R.string.connect_backup_code_mismatch);
                }
            } else {
                isValid = true;
            }
        }

        binding.connectBackupCodeErrorMessage.setVisibility(TextUtils.isEmpty(errorText) ? View.GONE : View.VISIBLE);
        binding.connectBackupCodeErrorMessage.setText(errorText);
        enableContinueButton(isValid);
    }

    private void handleNotMeButtonPressed(){
        personalIdSessionData.setAccountExists(false);
        clearBackupCodeFields();
        configureUiByMode();
    }

    private void enableContinueButton(boolean isEnable) {
        binding.connectBackupCodeButton.setEnabled(isEnable);
    }

    private void handleBackupCodeSubmission() {
        FirebaseAnalyticsUtil.reportPersonalIDContinueClicked(this.getClass().getSimpleName(), null);
        if (isRecovery) {
            confirmBackupCode();
        } else {
            personalIdSessionData.setBackupCode(
                    binding.backupCodeView.getCodeValue());
            if (ReleaseToggleHelper.INSTANCE.isEmailOtpVerificationActive(personalIdSessionData)) {
                navigateToEmail();
            } else {
                navigateToPhoto();
            }
        }
    }

    private void confirmBackupCode() {
        clearError();
        enableContinueButton(false);
        String backupCode = binding.backupCodeView.getCodeValue();

        new PersonalIdApiHandler<PersonalIdSessionData>() {
            @Override
            public void onSuccess(PersonalIdSessionData sessionData) {
                if (sessionData.getDbKey() != null) {
                    handleConfirmBackupCodeSuccess();
                } else if (sessionData.getAttemptsLeft() != null && sessionData.getAttemptsLeft() > 0) {
                    handleFailedBackupCodeAttempt();
                }
            }

            @Override
            public void onFailure(PersonalIdOrConnectApiErrorCodes failureCode, Throwable t) {
                if (handleCommonSignupFailures(failureCode)) {
                    return;
                }
                showError(PersonalIdOrConnectApiErrorHandler.handle(requireActivity(), failureCode, t));
                if (failureCode.shouldAllowRetry()) {
                    enableContinueButton(true);
                }
            }
        }.confirmBackupCode(activity, backupCode, personalIdSessionData);
    }

    private void handleConfirmBackupCodeSuccess() {
        if (personalIdSessionData.getEmail() == null &&
                ReleaseToggleHelper.INSTANCE.isEmailOtpVerificationActive(personalIdSessionData)) {
            navigateToEmail();
        } else {
            PersonalIdRecoveryCompleter.finalizeAccountRecovery(requireActivity(), personalIdSessionData);
            navigateToSuccess();
        }
    }

    private void clearError() {
        binding.connectBackupCodeErrorMessage.setVisibility(View.GONE);
        binding.connectBackupCodeErrorMessage.setText("");
    }

    private void showError(String message) {
        binding.connectBackupCodeErrorMessage.setVisibility(View.VISIBLE);
        binding.connectBackupCodeErrorMessage.setText(message);
    }

    private void handleFailedBackupCodeAttempt() {
        logRecoveryFailureResult();
        clearBackupCodeFields();
        navigateWithMessage(getString(R.string.connect_backup_fail_title),
                getString(R.string.personalid_wrong_backup_message, personalIdSessionData.getAttemptsLeft()),
                ConnectConstants.PERSONALID_RECOVERY_WRONG_BACKUPCODE);
    }

    private void navigateToEmail() {
        PersonalIdPreferences.setLastEmailOfferDate(activity, new Date());
        EmailWorkFlow emailWorkFlow = isRecovery ? EmailWorkFlow.RECOVERY : EmailWorkFlow.REGISTRATION;
        NavDirections action = PersonalIdBackupCodeFragmentDirections
                .actionPersonalidBackupcodeToPersonalidEmail(emailWorkFlow);
        Navigation.findNavController(binding.getRoot()).navigate(action);
    }

    private void logRecoveryFailureResult() {
        FirebaseAnalyticsUtil.reportPersonalIdAccountRecovered(false, AnalyticsParamValue.CCC_RECOVERY_METHOD_BACKUPCODE);
    }

    private void navigateWithMessage(String titleRes, String msgRes, int phase) {
        navigateToMessageDisplay(titleRes, msgRes, false, phase, R.string.ok);
    }

    private void navigateToPhoto() {
        Navigation.findNavController(binding.getRoot())
                .navigate(PersonalIdBackupCodeFragmentDirections
                        .actionPersonalidBackupcodeToPersonalidPhotoCapture());
    }

    private void navigateToSuccess() {
        navigateWithMessage(
                getString(R.string.connect_recovery_success_title),
                getString(R.string.connect_recovery_success_message),
                ConnectConstants.PERSONALID_RECOVERY_SUCCESS);
    }

    @Override
    protected void navigateToMessageDisplay(@NotNull String title, @Nullable String message, boolean isCancellable,
            int phase, int buttonText) {
        NavDirections navDirections = PersonalIdBackupCodeFragmentDirections
                .actionPersonalidBackupcodeToPersonalidMessage(title, message, phase,
                        getString(buttonText), null)
                .setIsCancellable(isCancellable);
        Navigation.findNavController(binding.getRoot()).navigate(navDirections);
    }
}

File: app/src/org/commcare/fragments/personalId/PersonalIdBackupCodeFragment.kt
package org.commcare.fragments.personalId

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavDirections
import androidx.navigation.findNavController
import org.commcare.activities.connect.viewmodel.PersonalIdSessionDataViewModel
import org.commcare.android.database.connect.models.PersonalIdSessionData
import org.commcare.connect.ConnectConstants
import org.commcare.connect.ReleaseToggleHelper
import org.commcare.connect.network.base.BaseApiHandler.PersonalIdOrConnectApiErrorCodes
import org.commcare.connect.network.base.PersonalIdOrConnectApiErrorHandler
import org.commcare.connect.network.personalId.PersonalIdApiHandler
import org.commcare.dalvik.R
import org.commcare.google.services.analytics.AnalyticsParamValue
import org.commcare.google.services.analytics.FirebaseAnalyticsUtil
import org.commcare.personalId.PersonalIdRecoveryCompleter
import org.commcare.personalId.PersonalIdUserPreferences
import org.commcare.utils.MediaUtil
import java.util.Date

class PersonalIdBackupCodeFragment : BasePersonalIdBackupCodeFragment() {
    private lateinit var personalIdSessionData: PersonalIdSessionData
    private var isRecovery = false

    override fun onResume() {
        super.onResume()
        validateBackupCodeAndEnableContinue()
    }

    override fun initData() {
        personalIdSessionData =
            ViewModelProvider(requireActivity())[PersonalIdSessionDataViewModel::class.java]
                .personalIdSessionData
    }

    override fun setUpView() {
        isRecovery = personalIdSessionData.accountExists == true
        if (isRecovery) {
            setUpInitialState(
                titleResId = R.string.connect_backup_code_title_confirm,
                showConfirmCode = false,
                subtitle = getString(R.string.connect_backup_code_message),
            )
            binding.recoveryCodeTilte.setText(R.string.connect_backup_code_message_title)
            binding.welcomeBackLayout.visibility = View.VISIBLE
            setUserNameAndPhoto()
            binding.personalidForgotBackupCode.visibility =
                if (!personalIdSessionData.email.isNullOrEmpty()) View.VISIBLE else View.GONE
        } else {
            setUpInitialState(
                titleResId = R.string.connect_backup_code_title_set,
                showConfirmCode = true,
                subtitle = getString(R.string.connect_backup_code_remember, BACKUP_CODE_LENGTH),
            )
        }
    }

    private fun setUserNameAndPhoto() {
        binding.welcomeBack.text = getString(R.string.personalid_welcome_back_msg, personalIdSessionData.userName)
        val photoBase64 = personalIdSessionData.photoBase64
        if (!photoBase64.isNullOrEmpty()) {
            binding.userPhoto.setImageBitmap(MediaUtil.decodeBase64EncodedBitmap(photoBase64))
        }
    }

    override fun setupListeners() {
        super.setupListeners()
        binding.backupCodeView.setCodeCompleteListener { if (isRecovery) submitIfEnabled() }
        binding.confirmCodeView.setCodeCompleteListener { submitIfEnabled() }
        binding.notMeButton.setOnClickListener { handleNotMeButtonPressed() }
        binding.personalidForgotBackupCode.setOnClickListener { handleForgotBackupCode() }
    }

    private fun handleNotMeButtonPressed() {
        personalIdSessionData.accountExists = false
        clearBackupCodeFields()
        setUpView()
    }

    private fun handleForgotBackupCode() {
        val bundle =
            Bundle().apply {
                putString("email", personalIdSessionData.email!!)
                putBoolean("masked", true)
                putSerializable("workflow", EmailWorkFlow.BACKUP_CODE_RECOVERY_SIGN_IN)
            }
        binding.root
            .findNavController()
            .navigate(R.id.action_personalid_backup_code_to_send_email_otp, bundle)
    }

    override fun handleBackupCodeSubmission() {
        FirebaseAnalyticsUtil.reportPersonalIDContinueClicked(
            javaClass.simpleName,
            null,
            PersonalIdWorkflow.CONFIGURATION,
        )
        personalIdSessionData.backupCode = binding.backupCodeView.codeValue
        if (isRecovery) {
            confirmBackupCode()
        } else {
            if (ReleaseToggleHelper.isEmailOtpVerificationActive(personalIdSessionData)) {
                navigateToEmail()
            } else {
                navigateToPhoto()
            }
        }
    }

    private fun confirmBackupCode() {
        clearError()
        enableContinueButton(false)
        val backupCode = binding.backupCodeView.codeValue

        object : PersonalIdApiHandler<PersonalIdSessionData>() {
            override fun onSuccess(sessionData: PersonalIdSessionData) {
                if (sessionData.dbKey != null) {
                    handleConfirmBackupCodeSuccess()
                } else if ((sessionData.attemptsLeft ?: 0) > 0) {
                    handleFailedBackupCodeAttempt()
                }
            }

            override fun onFailure(
                failureCode: PersonalIdOrConnectApiErrorCodes,
                t: Throwable?,
            ) {
                if (handleCommonSignupFailures(failureCode)) {
                    return
                }
                showError(PersonalIdOrConnectApiErrorHandler.handle(requireActivity(), failureCode, t))
                if (failureCode.shouldAllowRetry()) {
                    enableContinueButton(true)
                }
            }
        }.confirmBackupCode(activity, backupCode, personalIdSessionData)
    }

    private fun handleConfirmBackupCodeSuccess() {
        if (personalIdSessionData.email == null &&
            ReleaseToggleHelper.isEmailOtpVerificationActive(personalIdSessionData)
        ) {
            navigateToEmail()
        } else {
            PersonalIdRecoveryCompleter.finalizeAccountRecovery(requireActivity(), personalIdSessionData)
            navigateToSuccess()
        }
    }

    private fun handleFailedBackupCodeAttempt() {
        logRecoveryFailureResult()
        clearBackupCodeFields()
        navigateWithMessage(
            getString(R.string.connect_backup_fail_title),
            getString(R.string.personalid_wrong_backup_message, personalIdSessionData.attemptsLeft),
            ConnectConstants.PERSONALID_RECOVERY_WRONG_BACKUPCODE,
        )
    }

    private fun navigateToEmail() {
        PersonalIdUserPreferences.setLastEmailOfferDate(Date())
        val emailWorkFlow = if (isRecovery) EmailWorkFlow.RECOVERY else EmailWorkFlow.REGISTRATION
        navigate(
            PersonalIdBackupCodeFragmentDirections
                .actionPersonalidBackupcodeToPersonalidEmail(emailWorkFlow),
        )
    }

    private fun logRecoveryFailureResult() {
        FirebaseAnalyticsUtil.reportPersonalIdAccountRecovered(
            false,
            AnalyticsParamValue.CCC_RECOVERY_METHOD_BACKUPCODE,
        )
    }

    private fun navigateWithMessage(
        title: String,
        message: String,
        phase: Int,
    ) {
        navigateToMessageDisplay(title, message, false, phase, R.string.ok)
    }

    private fun navigateToPhoto() {
        navigate(PersonalIdBackupCodeFragmentDirections.actionPersonalidBackupcodeToPersonalidPhotoCapture())
    }

    private fun navigateToSuccess() {
        navigateWithMessage(
            getString(R.string.connect_recovery_success_title),
            getString(R.string.connect_recovery_success_message),
            ConnectConstants.PERSONALID_RECOVERY_SUCCESS,
        )
    }

    override fun navigateToMessageDisplay(
        title: String,
        message: String?,
        isCancellable: Boolean,
        phase: Int,
        buttonText: Int,
    ) {
        navigate(
            PersonalIdBackupCodeFragmentDirections
                .actionPersonalidBackupcodeToPersonalidMessage(
                    title,
                    message.orEmpty(),
                    phase,
                    getString(buttonText),
                    null,
                ).setIsCancellable(isCancellable),
        )
    }

    private fun navigate(directions: NavDirections) = binding.root.findNavController().navigate(directions)
}
